### PR TITLE
 Implement Reactor API/GRPC server side binding

### PR DESCRIPTION
**Pull request by tbak**

### CODE REVIEW

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/common/reactor/server/ServerStreamingMethodHandler.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+                        logger.warn("Subscriber threw error in onError handler", ex);
+                    }
+                },
+                () -> {
+                    try {
+                        responseObserver.onCompleted();
+                    } catch (Exception ex) {
+                        logger.warn("Subscriber threw error in onCompleted handler. Retrying with onError", ex);
+                        try {
+                            responseObserver.onError(ex);
```
Comment (fabiokung) [inline titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/common/reactor/server/ServerStreamingMethodHandler.java]: I think this breaks the `StreamObserver` contract:

```
  /**
   * Receives a notification of successful stream completion.
   *
   * <p>May only be called once and if called it must be the last method called. In particular if an
   * exception is thrown by an implementation of {@code onCompleted} no further calls to any method
   * are allowed.
   */
  void onCompleted();
```

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/common/reactor/server/ServerStreamingMethodHandler.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+    @Override
+    public void invoke(REQ request, StreamObserver<RESP> responseObserver) {
+        super.invoke(request, responseObserver);
+    }
+
+    @Override
+    Disposable handleResult(Publisher<RESP> result, StreamObserver<RESP> responseObserver) {
+        return ((Flux<RESP>) result).subscribe(
+                responseObserver::onNext,
```
Comment (fabiokung) [inline titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/common/reactor/server/ServerStreamingMethodHandler.java]: we need to handle exceptions here and terminate the stream calling `onError()`. From the `StreamObserver` contract:

```
  /**
   * Receives a value from the stream.
   *
   * <p>Can be called many times but is never called after {@link #onError(Throwable)} or {@link
   * #onCompleted()} are called.
   *
   * <p>Unary calls must invoke onNext at most once.  Clients may invoke onNext at most once for
   * server streaming calls, but may receive many onNext callbacks.  Servers may invoke onNext at
   * most once for client streaming calls, but may receive many onNext callbacks.
   *
   * <p>If an exception is thrown by an implementation the caller is expected to terminate the
   * stream by calling {@link #onError(Throwable)} with the caught exception prior to
   * propagating it.
   *
   * @param value the value passed to the stream
   */
  void onNext(V value);
```

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/common/reactor/server/ServerStreamingMethodHandler.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    @Override
+    public void invoke(REQ request, StreamObserver<RESP> responseObserver) {
+        super.invoke(request, responseObserver);
+    }
+
+    @Override
+    Disposable handleResult(Publisher<RESP> result, StreamObserver<RESP> responseObserver) {
+        return ((Flux<RESP>) result).subscribe(
```
Comment (fabiokung) [inline titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/common/reactor/server/ServerStreamingMethodHandler.java]: `Flux.from(result)` is better than the explicit cast here

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/common/reactor/server/UnaryMethodHandler.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+                    },
+                    () -> {
+                        try {
+                            // Void must be mapped to GRPC/Empty value.
+                            responseObserver.onNext((RESP) Empty.getDefaultInstance());
+                            responseObserver.onCompleted();
+                        } catch (Exception ex) {
+                            logger.warn("Subscriber threw error in onCompleted handler. Retrying with onError", ex);
+                            try {
+                                responseObserver.onError(ex);
```
Comment (fabiokung) [inline titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/common/reactor/server/UnaryMethodHandler.java]: same comment about breaking the `StreamObserver` contract here

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/common/reactor/server/UnaryMethodHandler.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+                            } catch (Exception ignore) {
+                            }
+                        }
+                    }
+            );
+        } else {
+            disposable = monoResult.subscribe(
+                    responseObserver::onNext,
```
Comment (fabiokung) [inline titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/common/reactor/server/UnaryMethodHandler.java]: similar comments from `ServerStreamingMethodHandler`

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/common/reactor/GrpcToReactorServerFactoryComponent.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+package com.netflix.titus.runtime.connector.common.reactor;
+
+import com.netflix.archaius.api.annotations.Configuration;
+import com.netflix.titus.runtime.connector.common.reactor.server.DefaultGrpcToReactorServerFactory;
+import com.netflix.titus.runtime.endpoint.metadata.CallMetadataResolver;
+import org.springframework.context.annotation.Bean;
+
+@Configuration
```
[the remark this answers is not in the document]
Comment (tbak) [reply titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/common/reactor/GrpcToReactorServerFactoryComponent.java]: Good catch.

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/common/reactor/server/ServerStreamingMethodHandler.java (hunk 1 above)
Comment (tbak) [reply titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/common/reactor/server/ServerStreamingMethodHandler.java]: I will keep the warn, but remove the subsequent onError.

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/common/reactor/server/ServerStreamingMethodHandler.java (hunk 2 above)
Comment (tbak) [reply titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/common/reactor/server/ServerStreamingMethodHandler.java]: Fixed

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/common/reactor/server/ServerStreamingMethodHandler.java (hunk 3 above)
Comment (tbak) [reply titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/common/reactor/server/ServerStreamingMethodHandler.java]: Nice. Changed

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/common/reactor/server/UnaryMethodHandler.java (hunk 5 above)
Comment (tbak) [reply titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/common/reactor/server/UnaryMethodHandler.java]: Fixed

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/common/reactor/server/UnaryMethodHandler.java (hunk 4 above)
Comment (tbak) [reply titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/common/reactor/server/UnaryMethodHandler.java]: Fixed

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/common/reactor/server/ServerStreamingMethodHandler.java [hunk 7]
```diff
@@ -42,24 +42,31 @@ public void invoke(REQ request, StreamObserver<RESP> responseObserver) {
...
-        return ((Flux<RESP>) result).subscribe(
-                responseObserver::onNext,
+        return Flux.from(result).subscribe(
+                value -> {
+                    try {
+                        responseObserver.onNext(value);
+                    } catch (Exception e) {
+                        logger.warn("Subscriber threw error in onNext handler", e);
+                        try {
+                            responseObserver.onError(e);
```
[the remark this answers is not in the document]
Comment (tbak) [reply titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/common/reactor/server/ServerStreamingMethodHandler.java]: This should be handled by `ServerCallStreamObserver.setOnCancelHandler`. But I agree that the current implementation is not very elegant, and if the callback does not happen, we may continue emitting values to a closed channel.

**Pull request merged**